metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
 	}
 
 	families := []*dto.MetricFamily{}
-	for {
-		mf := &dto.MetricFamily{}
-		err = decoder.Decode(mf)
-		if err != nil {
-			if err == io.EOF {
-				break
+	var wg sync.WaitGroup
+	var decoderMutex = &sync.Mutex{}
+	var familiesMutex = &sync.Mutex{}
+	var done = false
+	var count = 0
+	for !done && count < maxGetFamiliesSubFuns {
+		wg.Add(1)
+		count++
+		go func(id int) {
+			for !done {
+				mf := &dto.MetricFamily{}
+				decoderMutex.Lock()
+				err = decoder.Decode(mf)
+				decoderMutex.Unlock()
+				if err != nil {
+					if err == io.EOF {
+						done = true
+					}
+				} else {
+					if familyPrefix == nil {
+						familiesMutex.Lock()
+						families = append(families, mf)
+						familiesMutex.Unlock()
+					} else {
+						found := false
+						for _, prefix := range familyPrefix {
+							if strings.HasPrefix(mf.GetName(), prefix) {
+								familiesMutex.Lock()
+								families = append(families, mf)
+								familiesMutex.Unlock()
+								found = true
+								break
+							}
+						}
+						if !found {
+							mf.Reset()
+						}
+					}
+				}
 			}
-		} else {
-			families = append(families, mf)
-		}
+			wg.Done()
```

Comment:
you can change that do **defer wg.Done()** and put it in the beginning of the routine and then have multiple return points. (i am assuming all the indentation of this routine and avoiding of returns is done so that you hit this line.)